source/simulator/introspection.cc

```diff
@@ -166,6 +166,90 @@ namespace aspect
   }
 
 
+  /**
+   * This is an internal deal.II function stolen from dof_tools.cc
+   */
+  template <int dim, int spacedim>
+  std::vector<unsigned char>
+  get_local_component_association (const FiniteElement<dim,spacedim>  &fe,
+                                   const ComponentMask        &component_mask)
+  {
+    std::vector<unsigned char> local_component_association (fe.dofs_per_cell,
+                                                            (unsigned char)(-1));
+
+    // compute the component each local dof belongs to.
+    // if the shape function is primitive, then this
+    // is simple and we can just associate it with
+    // what system_to_component_index gives us
+    for (unsigned int i=0; i<fe.dofs_per_cell; ++i)
+      if (fe.is_primitive(i))
```

Comment:
are we ever using non-primitive elements? we could avoid a lot of complexity if we assume that